src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+        // tricky and not worth the effort, so we don't do it.
+
+        // Figure out the layer this operation should go into.
+        LayerId layerToInsertInto = this->FindLayerToInsertOperationInto(secondGroup[0], opDuration);
+        for (long i = 1; i < nSecondGroup && layerToInsertInto != INVALID; i++)
+        {
+            layerToInsertInto =
+                max(layerToInsertInto, this->FindLayerToInsertOperationInto(secondGroup[i], opDuration));
+        }
+        for (long i = 0; i < nFirstGroup && layerToInsertInto != INVALID; i++)
+        {
+            layerToInsertInto = max(layerToInsertInto, this->FindLayerToInsertOperationInto(firstGroup[i], opDuration));
+        }
+        if (layerToInsertInto == INVALID)
+        {
+            layerToInsertInto = this->CreateNewLayer(opDuration);
+        }
+
+        // Add the operation into the layer.
+        this->AddOperationToLayer(id, layerToInsertInto);
+
+        // Update the state of the involved qubits.
+        for (long i = 0; i < nFirstGroup; i++)
+        {
+            this->UpdateQubitState(firstGroup[i], layerToInsertInto, opDuration);
+        }
+        for (long i = 0; i < nSecondGroup; i++)
+        {
+            this->UpdateQubitState(secondGroup[i], layerToInsertInto, opDuration);
+        }
+
+        return layerToInsertInto;
+    }
+
+    LayerId CTracer::InjectGlobalBarrier(OpId id, Duration duration)
+    {
+        LayerId layer = this->CreateNewLayer(duration);
+        this->metricsByLayer[layer].barrierId = id;
+        this->globalBarrier = layer;
+        return layer;
+    }
+
+    Result CTracer::TraceSingleQubitMeasurement(OpId id, Duration duration, Qubit target)
+    {
+        LayerId layerId = this->TraceSingleQubitOp(id, duration, target);
+        return reinterpret_cast<Result>(layerId);
+    }
+
+    Result CTracer::TraceMultiQubitMeasurement(OpId id, Duration duration, long nTargets, Qubit* targets)
+    {
+        LayerId layerId = this->TraceMultiQubitOp(id, duration, 0, nullptr, nTargets, targets);
+        return reinterpret_cast<Result>(layerId);
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::PrintLayerMetrics
+    //------------------------------------------------------------------------------------------------------------------
+    static std::string GetOperationName(OpId opId, const std::unordered_map<OpId, std::string>& opNames)
+    {
+        if (opId < 0)
+        {
+            return "";
+        }
+
+        auto nameIt = opNames.find(opId);
+        return nameIt == opNames.end() ? std::to_string(opId) : nameIt->second;
+    }
+    void CTracer::PrintLayerMetrics(std::ostream& out, const std::string& separator, bool printZeroMetrics) const
+    {
+        // Sort the operations by id so the output is deterministic.
+        std::set<OpId> seenOpsOrederedById(this->seenOps.begin(), this->seenOps.end());

```

Comment:
Minor optimization:
I see that `this->seenOps` is only appended to, then iterated through in the sorted order of unique elements. I'm afraid the `unordered_set` can be expensive. Consider 
```c++
std::vector<OpId> seenOps;   // Instead of unordered_set
this->seenOps.push_back(id);   // Instead of this->seenOps.insert(id);

// Instead of `std::set<OpId> seenOpsOrederedById()`:
std::sort(this->seenOps.begin(), this->seenOps.end());
.. = std::unique(this->seenOps.begin(), this->seenOps.end());
```